array.go

```diff
@@ -861,3 +861,29 @@ func countElement(array []interface{}, element interface{}) int {
 	}
 	return count
 }
+
+// Every runs the passed function on all the Elements
+// in the array and returns a new Array instance.
+//
+// Example:
+//
+//	array := NewArray(t, []interface{}{"foo", "bar"})
+//	newArray := array.Every(func(idx int, val *httpexpect.Value) {
+//		value.String().NotEmpty()
+//	})
+func (a *Array) Every(fn func(index int, value *Value)) *Array {
+	newArrayValue := []interface{}{}
+
+	for idx, val := range a.value {
+		valueChain := a.chain.clone()
+		valueChain.enter("Every[%d]", val)
+
+		newValue := newValue(valueChain, val)
+
+		fn(idx, newValue)
+
+		newArrayValue = append(newArrayValue, newValue.value)
+	}
+
+	return newArray(a.chain, newArrayValue)
+}
```

Comment:
The idea in the issue was a bit different: apply some *check* to every element, and return the original array (unmodified).

So that when you write:

```go
array.Every( .... )
```

you're saying: every element of the array should satisfy given condition.

And this code implement Every as a transformation - it assumes that the function modifies value, and returns a new array with modified values, instead of original array.

This is a legit feature, but there are two issues:

* First, for such function a better name would be Transform or Map (better Transform to avoid confusion with go maps)

* Second, httpexpect structs like Array, Object, etc are immutable; they have no public methods for changing value (note that in your test you're modifying private fields, which is not possible outside of httpexpect package); so Transform's function argument would have different signature - since it can't modify value, it would return a new value instead

So, I think `Every` should be implemented according to original idea, and `Transform` should be a separate feature.

Next, look, we have this issue: #169

It describes another function, `Filter`. And I think instead of implementing both `Filter` and `Transform`, we can upgrade description for `Filter` and make it looking like this:

```go
newArray := array.Filter(func(index int, value *httpexpect.Value) (*httpexpect.Value, bool) {
    // make assertions
    value.String().NotEmpty()

    // return boolean
    return v, true
})
```

This way, it will combine both features in one method: filtering-out values (if desired) and transforming values (if desired).

What do you think?

Reply:
In the issue description for `Every`, I forgot one important detail: if assertion inside function failed, the original object should be marked as failed as well. Sorry for inconvenience.

To achieve this, you can use `setFailCallback` as described in #168.

Reply:
We should mention two important facts in documentation comments:

* if assertion inside function is failed, the original object is marked as failed
* if assertion inside function is failed, Every will anyway execute function for every value, instead of stopping after first failure

Reply:
Understood. This was a misunderstanding on my end. I think now that I have some context of what `Every` should do. The original Idea makes sense.

Although for `Filter` I think we can keep the original idea and create separate function `Transform` which allows users to manipulate the elements in the array/slice and since the httpexpect objects are immutable we can return a new array. 
```

newArray := array.Transform(func(index int, value *httpexpect.Value) *httpexpect.Value {
    // make assertions
    value.String().NotEmpty()

    // return boolean
    return v
})
```
Something like this. 

In this way we can clearly indicate that Filter function is only for filtering and cannot modify or manipulate the values
whereas Transform is just to modify the value. In this way the user can chain the function.

```
newArray := array.Filter().Transform()
```

Same can be done on the object. 

Does this seem like a good idea?

Reply:
Cool.
Agree, seems that separate Filter and Transform functions would be more intuitive.

Reply:
Hey @gavv would you mind looking into this new function. If everything's alright I will write the test cases for this.

```go
func (a *Array) Every(fn func(index int, value *Value)) *Array {
	a.chain.enter("Every()")
	defer a.chain.leave()

	if fn == nil {
		a.chain.fail(AssertionFailure{
			Type: AssertNotNil,
			Errors: []error{
				errors.New("expected function, got nil"),
			},
		})
		return a
	}

	chainFailure := false

	for idx, val := range a.value {
		valueChain := a.chain.clone()
		valueChain.enter("Every[%d]", val)

		valueChain.setFatal(false)
		valueChain.setFailCallback(func() {
			chainFailure = true
		})

		fn(idx, newValue(valueChain, val))
	}

	if chainFailure {
		a.chain.fail(AssertionFailure{
			Type: AssertNotValid,
		})
	}

	return a
}

```